src/ring_theory/ore_localization/ore_set.lean

```diff
@@ -0,0 +1,112 @@
+
+/-- Common factors on the left can be turned into common factors on the right, a weak form of
+cancellability. -/
+def ore_left_cancel (r₁ r₂ : R) (s : S) (h : ↑s * r₁ = s * r₂) : {s' : S // r₁ * s' = r₂ * s'} :=
+ore_set.ore_left_cancel r₁ r₂ s h
+
+/-- The common right factor of an Ore set's weak cancellability property. -/
+def ore_left_cancel_factor (r₁ r₂ : R) (s : S) (h : (s : R) * r₁ = s * r₂) : S :=
+(ore_left_cancel r₁ r₂ s h).1
+
+/-- The defining equality of an Ore set's weak cancellability. -/
+lemma ore_left_cancel_factor_eq (r₁ r₂ : R) (s : S) (h : (s : R) * r₁ = s * r₂) :
+  r₁ * (ore_left_cancel_factor r₁ r₂ s h) = r₂ * (ore_left_cancel_factor r₁ r₂ s h) :=
+(ore_left_cancel r₁ r₂ s h).2
+
+/-- The Ore numerator of a fraction. -/
+def ore_num (r : R) (s : S) : R := ore_set.ore_num r s
+
+/-- The Ore denominator of a fraction. -/
+def ore_denom (r : R) (s : S) : S := ore_set.ore_denom r s
+
+/-- The Ore condition of a fraction, expressed in terms of `ore_num` and `ore_denom`. -/
+lemma ore_eq (r : R) (s : S) : r * (ore_denom r s) = s * (ore_num r s) := ore_set.ore_eq r s
+
+/-- The Ore condition bundled in a sigma type. This is useful in situations where we want to obtain
+both witnesses and the condition for a given fraction. -/
+def ore_condition (r : R) (s : S) : Σ' r' : R, Σ' s' : S, r * s' = s * r' :=
+⟨ore_num r s, ore_denom r s, ore_eq r s⟩
+
+/-- The trivial submonoid is an Ore set. -/
+instance ore_set_bot : ore_set (⊥ : submonoid R) :=
+{ ore_left_cancel := λ _ _ s h,
+    ⟨1, begin
+          rcases s with ⟨s, hs⟩, rw submonoid.mem_bot at hs,subst hs,
+          rw [set_like.coe_mk, one_mul, one_mul] at h, subst h
+        end⟩,
+  ore_num := λ r _, r,
+  ore_denom := λ _ _, 1,
+  ore_eq := λ _ s, by { rcases s with ⟨s, hs⟩, rw submonoid.mem_bot at hs, simp [hs] } }
+
+/-- Every submonoid of a commutative monoid is an Ore set. -/
+def ore_set_comm {R} [comm_monoid R] (S : submonoid R) : ore_set S :=
+{ ore_left_cancel := λ m n s h, ⟨s, by rw [mul_comm n s, mul_comm m s, h]⟩,
+  ore_num := λ r _, r,
+  ore_denom := λ _ s, s,
+  ore_eq := λ r s, by rw mul_comm }
+
+attribute [instance, priority 100] ore_set_comm
+
+end monoid
+
+section ring
+
+variables {R : Type*} [ring R] [no_zero_divisors R] {S : submonoid R}
+
+/-- In rings without zero divisors, the first (cancellability) condition is always fulfilled,
+it suffices to give a prove for the ore condition itself. -/
+def ore_set_of_no_zero_divisors
+  (ore_num : R → S → R)
+  (ore_denom : R → S → S)
+  {ore_eq : ∀ (r : R) (s : S), r * (ore_denom r s) = s * (ore_num r s)} :
+  ore_set S :=
+{ ore_left_cancel := λ r₁ r₂ s h,
+  begin
+    cases s with s hs,
+    use ⟨s, hs⟩,
+    have : s * (r₂ - r₁) = 0,
+    { rw [mul_sub], apply sub_eq_zero_of_eq, symmetry, assumption },
+    cases eq_zero_or_eq_zero_of_mul_eq_zero this with h0 h0,
+    { subst h0, simp },
+    { replace h0 := eq_of_sub_eq_zero h0, subst h0 }
+  end,
```

Comment:
Here's a shorter proof that's also definitionally equal to the other instance above:
```suggestion
{ ore_left_cancel := λ r₁ r₂ s h, ⟨s, begin
    letI : cancel_monoid_with_zero R := no_zero_divisors.to_cancel_monoid_with_zero,
    rw mul_eq_mul_right_iff,
    rw mul_eq_mul_left_iff at h,
    exact h,
  end⟩,
```

Reply:
Or even
```suggestion
{ ore_left_cancel := λ r₁ r₂ s h, ⟨s, begin
    letI : cancel_monoid_with_zero R := no_zero_divisors.to_cancel_monoid_with_zero,
    exact mul_eq_mul_right_iff.mpr (mul_eq_mul_left_iff.mp h)
  end⟩,
```

Reply:
Can you just assume `cancel_monoid_with_zero R` instead of `ring` + `no_zero_divisors`? We have the instance `[is_domain.to_cancel_monoid_with_zero](https://leanprover-community.github.io/mathlib_docs/algebra/ring/basic.html#is_domain.to_cancel_monoid_with_zero)` and `[is_domain.to_cancel_comm_monoid_with_zero](https://leanprover-community.github.io/mathlib_docs/algebra/ring/basic.html#is_domain.to_cancel_comm_monoid_with_zero)` and if we generalize `is_domain` to `no_zero_divisors`, then this instance will apply to `ring` + `no_zero_divisors`.

Reply:
Good point, added it!